File: crates/brainwires-mcp-server/src/handler.rs
use anyhow::Result;
use async_trait::async_trait;
use brainwires_mcp_client::{CallToolResult, InitializeParams, ServerCapabilities, ServerInfo};
use serde_json::Value;

use crate::connection::RequestContext;
use crate::registry::McpToolDef;

/// Trait for handling MCP protocol requests.
#[async_trait]
pub trait McpHandler: Send + Sync + 'static {
    /// Return server identification info.
    fn server_info(&self) -> ServerInfo;
    /// Return server capabilities.
    fn capabilities(&self) -> ServerCapabilities;
    /// List all available tools.
    fn list_tools(&self) -> Vec<McpToolDef>;
    /// Execute a tool call.
    async fn call_tool(
        &self,
        name: &str,
        args: Value,
        ctx: &RequestContext,
    ) -> Result<CallToolResult>;

    /// Called when the client sends an initialize request.
    async fn on_initialize(&self, _params: &InitializeParams) -> Result<()> {
        Ok(())
    }

    /// Called when the server is shutting down.
    async fn on_shutdown(&self) -> Result<()> {
        Ok(())
    }
}

File: deno/packages/mcp-server/handler.ts
/**
 * @module handler
 *
 * MCP request handler interface.
 * Equivalent to Rust's `McpHandler` trait.
 */

import type {
  CallToolResult,
  InitializeParams,
  ServerCapabilities,
  ServerInfo,
} from "@rullama/mcp-client";
import type { McpToolDef } from "./registry.ts";
import type { RequestContext } from "./server.ts";

/**
 * Interface for handling MCP protocol requests.
 * Equivalent to Rust `McpHandler` trait.
 */
export interface McpHandler {
  /** Return server identification info. */
  serverInfo(): ServerInfo;

  /** Return server capabilities. */
  capabilities(): ServerCapabilities;

  /** List all available tools. */
  listTools(): McpToolDef[];

  /** Execute a tool call. */
  callTool(
    name: string,
    args: Record<string, unknown>,
    ctx: RequestContext,
  ): Promise<CallToolResult>;

  /** Called when the client sends an initialize request. */
  onInitialize?(params: InitializeParams): Promise<void>;

  /** Called when the server is shutting down. */
  onShutdown?(): Promise<void>;
}
